core/quill.ts

```diff
@@ -917,11 +952,24 @@ function overload(
   return [index, length, formats, source];
 }
 
-function shiftRange(range, index, length, source?: EmitterSource) {
+function shiftRange(range: Range, change: Delta, source?: EmitterSource): Range;
+function shiftRange(
+  range: Range,
+  index: number,
+  length?: number,
+  source?: EmitterSource,
+): Range;
+function shiftRange(
+  range: Range,
+  index: number | Delta,
+  _length?: number | EmitterSource,
+  source?: EmitterSource,
+) {
+  const length = typeof _length === 'number' ? _length : 0;
   if (range == null) return null;
   let start;
   let end;
-  if (index && typeof index.transformPosition === 'function') {
+  if (index && index instanceof Delta) {
```

Comment:
We should avoid changing actual code. This brings a breaking change when users provide another Delta library or instance.

Reply:
I can change it back to check for the typeof the transformPosition function, but we _will_ at least need to add a check here that the `typeof index !== 'number'` (because `index` is either a number or a Delta at this point and we don't know which). Without a better type guard there, we get this error:
```
Property 'transformPosition' does not exist on type 'number | Delta'.
  Property 'transformPosition' does not exist on type 'number'. ts(2339)
```

I get the desire to avoid code changes and I did in fact try to avoid them wherever possible, but I erred on the side of adding better guards instead of just suppressing type errors. Our bug tracker at work reports tens of thousands of exceptions from Quill every day and part of my motivation with this PR was to find and fix the source of those 😄

Reply:
Yeah feel free to create issues so I can take a look at the errors. We could address them in separate PRs.